### PR TITLE
[custom_ops] Support TensorList inputs/outputs

### PR DESCRIPTION
**Pull request by zou3519**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #124092
* #124089
* #124071
* #124066
* #124065
* #124064
* #123937
* #124063
* #124062
* __->__ #123615

We add a `supports_tensorlist` decorator that gives an autograd.Function
the ability to handle TensorLists.

Test Plan:
- custom_op_db tests

### CODE REVIEW

**Comment by vadimkantorov** [comment]
Related: https://github.com/pytorch/pytorch/issues/100249 :)

File: torch/_library/autograd.py [hunk 1]
```diff
@@ -31,8 +32,95 @@ def backward(ctx, *grads):
...
+
+    cls.forward = new_forward
+    cls.backward = new_backward
+    cls.apply = new_apply
+    return cls
+
+
+import functools
+
+from ..utils import _pytree
```
Comment (albanD) [inline torch/_library/autograd.py:111]: Why is this at the bottom of the file?

File: torch/_library/autograd.py [hunk 2]
```diff
@@ -31,8 +32,95 @@ def backward(ctx, *grads):
...
+        Generated = supports_tensorlist(Generated)
+
     def autograd_impl(*args):
         result = Generated.apply(*args)  # type: ignore[attr-defined]
         return result
 
     return autograd_impl
+
+
+def supports_tensorlist(cls: Any) -> Any:
```
Comment (albanD) [inline torch/_library/autograd.py:49]: There is quite a bit of state here.
I think it would be good to have a few asserts to make sure that we only ever do apply -> fw -> bw -> apply -> fw -> bw -> etc
And that any double apply/bw/etc will raise a nice error saying this is not supported yet.

**Review by albanD — COMMENTED** [review]
Sounds fair. Small comment on more constraints to raise errors on weird behavior!

File: torch/_library/autograd.py (hunk 1 above)
Comment (zou3519) [reply torch/_library/autograd.py:111]: I hoped lintrunner would move it up but looks like it didn't

File: torch/_library/autograd.py (hunk 2 above)
Comment (zou3519) [reply torch/_library/autograd.py:49]: That's a good catch. Let me think about if we can support the double apply situation -- that is probably going to be common.

Comment (zou3519) [reply torch/_library/autograd.py:51]: Updated with nice error messages. I also expand the number of things we support: we can support double apply. The secret is to put the input_spec/output_spec onto the ctx object so that each of the backward have their own input_spec/output_spec

File: torch/testing/_internal/custom_op_db.py [hunk 3]
```diff
@@ -177,97 +177,87 @@ def sample_inputs_numpy_view_copy(opinfo, device, dtype, requires_grad, **kwargs
     result = make_arg(2, 3, 4, low=0.9, high=2)
     yield SampleInput(result, args=([2, 12],))
 
-@custom_ops.custom_op('_torch_testing::numpy_cat')
+@torch.library.custom_op('_torch_testing::numpy_cat', mutates_args=())
```
Comment (ezyang) [inline torch/testing/_internal/custom_op_db.py:180]: As best I can tell, the changes here are just... refactoring? You never use the new decorator here

Comment (zou3519) [reply torch/testing/_internal/custom_op_db.py:180]: We register an autograd.Function to the Autograd dispatch key. In this PR, we apply the supports_tensorlist to said autograd.Function. This op, using the new custom ops API, would not have passed the new tests without the new decorator.

File: test/test_custom_ops.py [hunk 4]
```diff
@@ -2137,6 +2137,61 @@ def _(x, y):
...
+        with self.assertRaisesRegex(NotImplementedError, "Function.backward directly"):
+            Stack.backward(None, xs)
+
+        # the recursive case
+        @torch._library.autograd.supports_tensorlist
+        class Foo(torch.autograd.Function):
+            @staticmethod
+            def forward(ctx, xs):
+                if len(xs) > 0:
+                    return Foo.apply(xs[1:])
```
Comment (albanD) [inline test/test_custom_ops.py:2182]: Is doing a Foo.apply() during the "backward()" call ok? I would expect it to be a more likely thing users will do than this one.

File: torch/_library/autograd.py [hunk 5]
```diff
@@ -31,8 +34,136 @@ def backward(ctx, *grads):
...
+    Regular autograd.Function has a constraint that it only directly supports autograd for
+    Tensors. Applying @supports_tensorlist enables an autograd.Function to support
+    autograd for List[Tensor] inputs and outputs.
+    """
+    # NB: All calls to the autograd.Function.apply shares these variables
+    # We assume that only one call to .apply happens at a time. This means that
+    # you cannot call the autograd.Function recursively (e.g. from its own forward).
+    input_spec: Optional[spec_t] = None
+    output_spec: Optional[spec_t] = None
+    result_is_tuple = None
```
Comment (albanD) [inline torch/_library/autograd.py:63]: I think you can simplify most of these by:
- Returning specs/is_tuple as output to the forward. Any non-Tensor will just be ignored.
- Use a single nonlocal boolean flag to detect re-entrancy and keep the error checking

File: torch/_library/autograd.py [hunk 6]
```diff
@@ -31,8 +34,136 @@ def backward(ctx, *grads):
...
+            return result[0]
+        return result
+
+    cls.forward = new_forward
+    cls.backward = new_backward
+    cls.apply = new_apply
+    return cls
+
+
+def not_list_of_tensor(tree):
```
Comment (albanD) [inline torch/_library/autograd.py:151]: I don't recall the full details, but what's the link between our TensorList and tuple(Tensor,...) ?
I guess we convert tuple(Tensor, ...) to TensorList on the way in. But do we return actual lists?
I'm ok with being very strict here but that might be surprising to some users.

File: test/test_custom_ops.py (hunk 4 above)
Comment (zou3519) [reply test/test_custom_ops.py:2182]: Yes, I can add a test for this

File: torch/_library/autograd.py (hunk 5 above)
Comment (zou3519) [reply torch/_library/autograd.py:63]: I'm not sure this is cleaner because:
- we still need a nonlocal input_spec. Because we need to ferry this information from the .apply to the .forward
- adding additional outputs to forward means that we need to consider them in backward, which makes the backward function more complicated.

But I'm happy to try if you think this is better.

File: torch/_library/autograd.py (hunk 6 above)
Comment (zou3519) [reply torch/_library/autograd.py:151]: For inputs, the pytorch dispatcher converts them to list
For outputs, the pytorch dispatcher converts tuples to list

I expect users to have trouble with the strictness (there are other problems, like not being able to pass None as the gradient for a TensorList), but I don't want to relax them in this PR (it's already getting complicated).

File: torch/_library/autograd.py (hunk 5 above)
Comment (albanD) [reply torch/_library/autograd.py:63]: I can buy that!

**Review by albanD — APPROVED** [review]
Ok sounds good!

**Review by albanD — APPROVED** [review]
